Wallet v2

ekzyis opened the pull request:
## Description

Close #1495 

This will:

- refactor vault
  - backend (#2092) :white_check_mark:
  - frontend (encryption and decryption) :construction:
- store send and receive config in separate tables (#2146) :white_check_mark: 
- support multiple receive and send protocols per wallet
  - backend (#2146) :white_check_mark:
  - frontend :construction: 
- separate wallets from protocols for better UI/UX/DX
  - backend (#2146)) :white_check_mark: 
  - frontend :construction: 
- create a list of wallets with the protocols they support
  - backend (#2146)) :white_check_mark: 
  - frontend :construction: 
- store all wallet logs on the server :construction:
- let users search for their wallet :construction: 

## Screenshots

tbd

## Additional Context

tbd

## Checklist

**Are your changes backwards compatible? Please answer below:**

tbd

**On a scale of 1-10 how well and how have you QA'd this change and any features it might affect? Please answer below:**

tbd

**For frontend changes: Tested on mobile, light and dark mode? Please answer below:**

tbd

**Did you introduce any new environment variables? If so, call them out explicitly here:**

tbd

ekzyis commented:
> ### Bug: Hook Accesses Undefined URL Parameters
> The `useSendRecvParam` and `useWalletProtocolParam` hooks directly access `params.slug[1]` and `params.slug[2]` respectively. This can cause a `TypeError` if `params.slug` is `undefined` or lacks sufficient elements for the current URL path.
> 

`params.slug` will never be undefined in this context.

But this reminded me that I wanted to refactor the hooks in wallets/client/components/forms.js. I think I went a bit overboard with small hooks. I think I can merge at least the ones that wrap `usePathname` or `useParams`.

ekzyis commented:
I rebased this PR on 17aada6d to fix conflicts but I will stop rebasing now because @huumn is reviewing

huumn commented:
Feel free to merge updates, but rebasing or forced pushes will make me grumpy. :D

huumn commented on wallets/client/components/passphrase.js:
```diff
@@ -0,0 +1,61 @@
+
+  const updateWalletEncryption = useWalletEncryptionUpdate()
+
+  useEffectOnce(() => {
+    async function updateKey () {
+      const { passphrase, key, hash } = await generateRandomKey()
+      await updateWalletEncryption({ key, hash, wallets })
+      setPassphrase(passphrase)
+    }
+    updateKey()
+  }, [])
```
`useEffect` in strict mode (which is only on in dev) runs twice, so you can detect when you're doing things you should not be doing in `useEffect`.

ekzyis replied:
Mhh yes, but I think strict mode was messing with development too much here.

But I don't remember if it broke something or it was just an unnecessary API call.
Should I not generate the new passphrase and run the migration in an `useEffect`?

> like doing something that isn't idempotent.

I guess I could generate the passphrase and run the wallet encryption in the `onShow` callback of `useShowPassphrase` and then just pass `passphrase` to the `Passphrase` component.

huumn replied:
Anything that doesn't require this hack, and isn't a hack itself, is probably better. I think that's why strict mode exists - prevent devs from doing stuff based on the component lifecycle that they shouldn't be doing.

ekzyis commented:
remaining TODOs:

- [x] wallets/README.md
- [ ] use `templateName` instead of `templateId`
- [ ] test client wallet migration without IndexedDB

huumn commented on prisma/migrations/20250702000001_wallet_v2/migration.sql:
```diff
@@ -0,0 +1,1126 @@
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER wallet_check_support
+    BEFORE INSERT OR UPDATE ON "WalletSendNWC"
+    FOR EACH ROW
+    EXECUTE FUNCTION wallet_check_support('SEND', 'NWC');
```
We'd also have to update these to be `AFTER` triggers, which might mess with some of the sequence manipulations, but idk.

Anyway, just something I saw while reviewing the docs.

ekzyis commented:
Regarding duplicate wallets during migration of local wallets:

I tested this with and without strict mode enabled, these are the results:

```
strict mode:     ..x.x.x..x..x.xx..x 
w/o strict mode: ...................
```
(x means duplicate wallet was created)

I think it doesn't always happen with strict mode because it depends on when the wallets query returns results.

But I think the results are quite clear: it happens because of strict mode, so it shouldn't happen in prod.

huumn commented:
> Regarding duplicate wallets during migration of local wallets:

This happens because migration is triggered as part of components loading, right? I wonder if there's a way to this kind of thing without hacks.

ekzyis commented:
> This happens because migration is triggered as part of components loading, right?

Yes

> I wonder if there's a way to this kind of thing without hacks.

The only guidance I found in the [React docs regarding this](https://react.dev/learn/synchronizing-with-effects#how-to-handle-the-effect-firing-twice-in-development) are the following sections:

- [Fetching data](https://react.dev/learn/synchronizing-with-effects#fetching-data) suggests to add a cleanup function but this doesn't help us because it still sends two requests, it just discards the second response
- [Sending analytics](https://react.dev/learn/synchronizing-with-effects#sending-analytics) suggests to ignore duplicate requests in dev because they won't happen in prod
- [Not an Effect: Buying a product](https://react.dev/learn/synchronizing-with-effects#not-an-effect-buying-a-product) suggests using a button instead of an effect

Since I don't think we want to make the migration depend on a button click, I think we're stuck with ignoring duplicate wallets during strict mode.

huumn commented:
> Since I don't think we want to make the migration depend on a button click, I think we're stuck with ignoring duplicate wallet migration in dev with strict mode enabled.

The advice in this section seems applicable: https://react.dev/learn/synchronizing-with-effects#not-an-effect-initializing-the-application

This might require refactoring migration away from being a hook though.

huumn commented:
It might not be worth it. I know a lot of the calls are hooks for code reusability.

ekzyis commented:
> It might not be worth it. I know a lot of the calls are hooks for code reusability.

Yeah I'd have to fetch `ME`, `WALLETS` and decrypt the wallets manually etc.

However, I wonder if I could mount a button and then click it once using this approach ... but I think that's also not worth it.

It's annoying in dev, but I am pretty sure this won't happen in prod.

ekzyis commented:
I refactored the wallet reducer in 8be0b30.

Instead of storing `page` + `loading` in the state, I noticed what I actually wanted was to have `status`. This meant I could also remove router events to "reset the page" when we left /wallets. I can just use a regular button to go from 'attach wallets' to the list of wallets instead of "changing the page."

I also fixed the flicker of 'attach wallets' -> 'unlock wallets' or 'attach wallets' -> list of wallets in 8be0b303 by adding a loading spinner while we're still determining what we want to show:

<img width="400" height="711" alt="localhost_3000_wallets(iPhone SE)" src="https://github.com/user-attachments/assets/9255d07a-a5ee-49b8-b781-b2ecac70c68b" />

With this refactor, I can now easily add another status when the wallets are unavailable instead of thinking in pages.

ekzyis commented:
I now show this message to the user if IndexedDB is not supported:

<img width="300" height="534" alt="localhost_3000_wallets(iPhone SE) (2)" src="https://github.com/user-attachments/assets/0c769532-67b8-47f2-b24b-6a70ba9661ce" />

ekzyis commented:
Okay, I’m done here now. As far as I am concerned, this can be merged!

ekzyis commented:
> On brand new accounts, I'm being asked to unlock wallets

Ohhh, good catch! There is a race condition in `useKeyInit`. The `useEffect` is called twice (not sure yet if that's also caused by strict mode), so if there's no key yet, we _might_ generate two random keys. The server saves the first key, whereas the client overwrites it with the second key, leading to a key mismatch.

huumn commented:
IME: let strict mode help you. Do not hack around it. 

It's particularly tricky with this stateful/sideeffect initialization stuff. We need to learn a decent pattern around it imo.
